source/Calamari.Aws/Integration/CloudFormation/CloudFormationObjectExtensions.cs

```diff
@@ -141,6 +142,64 @@ public static Task<CreateChangeSetResponse> CreateChangeSetAsync(this Func<IAmaz
             }
         }
 
+        /// <summary>
+        /// Gets all stack events, optionally filtered by a predicate
+        /// </summary>
+        /// <param name="predicate">The optional predicate used to filter events</param>
+        /// <returns>The stack events</returns>
+        public static async Task<List<Maybe<StackEvent>>> GetStackEvents(this Func<IAmazonCloudFormation> clientFactory,
+            StackArn stack,
+            Func<StackEvent, bool> predicate = null)
+        {
+            try
+            {
+                var stackEventResults = new List<Maybe<StackEvent>>();
+
+                var response = await clientFactory().DescribeStackEventsAsync(new DescribeStackEventsRequest { StackName = stack.Value });
+
+                var stackEvents = response?
+                    .StackEvents.Where(stackEvent => predicate == null || predicate(stackEvent))                    
+                    .ToList();
+
+                stackEventResults.AddRange(stackEvents.Select(s => s.AsSome()));
+
+                var nestedStackIds = stackEvents
+                    .Where(s => s.ResourceType == "AWS::CloudFormation::Stack" && s.PhysicalResourceId != s.StackId)
+                    .Select(s => s.PhysicalResourceId)
+                    .Distinct()
+                    .ToList();
+
+                foreach (var nestedStackId in nestedStackIds)
+                {
+                    var nestedStackEvents = await GetStackEvents(clientFactory, new StackArn(nestedStackId), predicate);
+                    if (nestedStackEvents.Any())
+                    {
+                        stackEventResults.AddRange(nestedStackEvents);
+                    }
+                }
+
+                return stackEventResults.OrderBy(s => s.SelectValueOr(e => e.Timestamp, DateTime.MinValue)).ToList();
+            }
+            catch (AmazonCloudFormationException ex) when (ex.ErrorCode == "AccessDenied")
+            {
+                throw new PermissionException(
+                    "The AWS account used to perform the operation does not have the required permissions to query the current state of the CloudFormation stack. " +
+                    "This step will complete without waiting for the stack to complete, and will not fail if the stack finishes in an error state.\n " +
+                    "Please ensure the current account has permission to perform action 'cloudformation:DescribeStackEvents'" +
+                    ex.Message);
+            }
+            catch (AmazonCloudFormationException ex) when (ex.ErrorCode == "ExpiredToken")
+            {
+                throw new PermissionException(
+                    "Security token has expired. Please increase the session duration and/or check that the system date and time are set correctly. " +
+                    ex.Message);
```

Comment:
Are we going to repeat the same error message here? I was looking to see what the original error message is and wondering if it is good enough?

Reply:
I'll check on the error message that comes from CloudFormation and see if it is good enough.

Reply:
This handling was added as part of https://github.com/OctopusDeploy/OctopusDeploy/pull/6187 (and associated Calamari PRs) where users were getting timeouts with non-helpful error messages because IAM role assumption was expiring. Based on the screenshots from the linked PR the message that comes in from the CloudFormation exception looks to be "The security token included in the request is expired", so it is somewhat duplicated with the first part of the custom message here, the second part gives some guidance on what to do about it. Do you think we need/want to make any changes here?

Reply:
Just remove "Security token has expired." ?

Reply:
We could probably put the message from AWS at the front, it would then read "The security token included in the request is expired. Please increase the session..." which seems to work pretty well. I wonder if this should also use the logging guidelines format and link to a known error code with docs, kind of like the "AWS-CLOUDFORMATION-ERROR-0001" error code.